[CIR] initial support for pointer-to-data-member type

Lancern opened the pull request:
This patch adds initial support for the pointer-to-data-member type. Specifically, this commit includes:

- New ops, types, and attributes:
  - Add a new type `!cir.member_ptr`
  - Add a new attribute `#cir.data_member_ptr`
  - Add a new operation `cir.get_indirect_member`
- CodeGen for pointer-to-data-member types and values
  - Lower C++ pointer-to-member type to `!cir.member_ptr`
  - Lower C++ expression `&C::D` to `cir.const(#cir.data_member_ptr)`
  - Lower C++ expression `c.*p` and `c->*p` to `cir.get_indirect_member`

This patch only includes an initial support. The following stuff related to pointer-to-member types are not supported yet:

- Pointer to member function;
- Conversion from `T Base::*` to `T Derived::*`;
- LLVMIR lowering.

Hopefully they will eventually be supported in later PRs :)

sitio-couto reviewed (commented):
@Lancern awsome! I left a few suggestions to improve testing and the codegen side.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenClass.cpp:
```diff
@@ -1476,3 +1476,63 @@ mlir::Value CIRGenFunction::getVTablePtr(SourceLocation Loc, Address This,
+  auto resultTy = builder.getPointerTo(memberPtrTy.getMemberTy());
+
+  auto op = builder.create<mlir::cir::GetIndirectMemberOp>(
+      getLoc(E->getSourceRange()), resultTy, base.getPointer(), memberPtr);
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -865,6 +865,27 @@ LValue CIRGenFunction::buildDeclRefLValue(const DeclRefExpr *E) {
+  else if (E->getOpcode() == BO_PtrMemI)
+    baseAddr = buildPointerWithAlignment(E->getLHS());
+  else
+    llvm_unreachable("unexpected binary operator opcode");
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -865,6 +865,27 @@ LValue CIRGenFunction::buildDeclRefLValue(const DeclRefExpr *E) {
+
+  auto memberPtr = buildScalarExpr(E->getRHS());
+
+  LValueBaseInfo baseInfo;
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExprConst.cpp:
```diff
@@ -1532,6 +1532,20 @@ mlir::Attribute ConstantEmitter::tryEmitPrivate(const APValue &Value,
+
+    const auto *fieldDecl = cast<FieldDecl>(memberDecl);
+    return mlir::cir::DataMemberPtrAttr::get(cirTy, cirTy.getClsTy(),
+                                             fieldDecl->getFieldIndex());
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -495,7 +495,11 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    if (llvm::isa<MemberPointerType>(E->getType())) {
+      const DeclRefExpr *DRE = cast<DeclRefExpr>(E->getSubExpr());
+      return CGF.buildMemberPtrConstant(DRE, E->getType());
+    }
```
Updated.

Lancern commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,89 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
 }
 
+//===----------------------------------------------------------------------===//
+// DataMemberPtrAttr definitions
```
What I'm thinking is that there's just too much differences between a ptr-to-data-member and a ptr-to-member-function, so it seems not reasonable enough to invent a general `#cir.member_ptr` attribute to represent both of them.

We can use a single `cir.member_ptr` type to represent both ptr-to-data-member type and ptr-to-function-member type because we can distinguish these two by simply testing whether the member type is a function type. However, while a ptr-to-data-member constant can be represented as a `(cls_ty, member_index)` pair, a ptr-to-function-member constant cannot be represented in a similar form. At least you need to pack a function pointer in it, and in the case of virtual member functions things are even more complex.

sitio-couto replied:
> At least you need to pack a function pointer in it, and in the case of virtual member functions things are even more complex.

Good point!

Lancern commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,89 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+}
+
+LogicalResult
+DataMemberPtrAttr::verify(function_ref<InFlightDiagnostic()> emitError, Type ty,
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExprConst.cpp:
```diff
@@ -1532,6 +1532,20 @@ mlir::Attribute ConstantEmitter::tryEmitPrivate(const APValue &Value,
     return buildArrayConstant(CGM, Desired, CommonElementType, NumElements,
                               Elts, typedFiller);
   }
+  case APValue::MemberPointer: {
```
Is it sufficient to add a comment saying that we're ignoring calls to `CGM.getCXXABI()` here?

Lancern replied:
Oh, seems we don't need CXXABI here in the first place? The CIR CodeGen of member pointers is ABI independent. The CXXABI is related only when the CIR is being lowered to LLVMIR.

sitio-couto replied:
This change is not necessarily to indicate that we **will** have to add a call to the CXXABI here but to identify paths that diverge from the original codegen! You might be right that we will never need a CXXABI call there, but it is still useful to know where they would be called when we decide how to tackle ABI-specific stuff.

Parity with the original codegen is one of our design principles. Even when we need to change the original codegen due to inherent differences between MLIR and LLVM IR, we leave notes clarifying why we must do so. For example:

https://github.com/llvm/clangir/blob/b42f8847437bc5e9a619db4dbb14137f588d661d/clang/lib/CIR/CodeGen/CIRGenExpr.cpp#L319-L325

Lancern replied:
Updated. Additionally found another 2 places that diverges from the original CodeGen on the usage of CXXABI.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenClass.cpp:
```diff
@@ -1476,3 +1476,63 @@ mlir::Value CIRGenFunction::getVTablePtr(SourceLocation Loc, Address This,
+CIRGenModule::getDynamicOffsetAlignment(clang::CharUnits actualBaseAlign,
+                                        const clang::CXXRecordDecl *baseDecl,
+                                        clang::CharUnits expectedTargetAlign) {
+  // If the base is an incomplete type (which is, alas, possible with
+  // member pointers), be pessimistic.
+  if (!baseDecl->isCompleteDefinition())
+    return std::min(actualBaseAlign, expectedTargetAlign);
```
Added tests for this.

sitio-couto reviewed (commented):
@Lancern thanks for all the changes! Just one more round of reviews.

bcardosolopes commented:
Thanks for adding support for pointers to data members! This is going to allows to build even more C++, way to go.

I believe we can use currently existing types and ops to support these, and no need to introduce new ops/attr/types. If you have a specific use case if you care about, can you please provide examples of how using this extra layer information helps here?

>   * Add a new type `!cir.member_ptr`

All ops that use/return these might as well return a `cir.ptr` to the actual final type.

>   * Add a new attribute `#cir.data_member_ptr`

The AST allows you to query the type of the struct element given you already have the index, so it can just be used directly.

>   * Add a new operation `cir.get_indirect_member`

Without a `!cir.member_ptr`, this one becomes unnecessary.

Lancern commented:
@bcardosolopes Thanks for your comment! But TBH I'm sorry I don't get your point exactly.

> All ops that use/return these might as well return a `cir.ptr` to the actual final type.

I don't clearly see how this is possible. A member pointer is totally different from an ordinary pointer. If we insist the reuse of the existing types instead of the new types, one possible way is to use the types compatible with the underlying ABI layout to represent a ptr-to-data-member. On most platforms, a ptr-to-data-member is simply represented by a signed integer which represents the possibly negative offset of the target field from the beginning of the containing object. So we might use a `!s32i` or a `!s64i` to represent a ptr-to-data-member. But this would make a ptr-to-data-member indistinguishable from a normal integer in the CIR, which might affect analysis that is interested on this. Also, this would make the CIR of `obj.*member` and `obj->*member` complex because you need to do a series of raw pointer arithmetic and casts.

Another possible way to represent the ptr-to-member type in CIR is to invent an imagined `cir.member` type and then `cir.ptr<cir.member>` represents a ptr-to-member. Are you suggesting this method? This method diverges from the C++ language since the language does not have a corresponding concept.

Lancern reviewed (commented):
Seems that I missed the null pointer value support for ptr-to-data-member. I'll check this and update the PR.

Lancern commented on clang/test/CIR/IR/invalid.cir:
```diff
@@ -781,3 +781,66 @@ module {
+!u32i = !cir.int<u, 32>
+!struct1 = !cir.struct<struct "Struct1" {!u16i, !u32i}>
+!struct2 = !cir.struct<struct "Struct2" {!u16i, !u32i}>
+module {
+  // expected-error@+2 {{type of a DataMemberPtrAttr must be MemberPtrType}}
+  // expected-error@+2 {{custom op 'cir.global' expected constant attribute to match type}}
+  cir.global external @invalid_type = #cir.data_member_ptr<!struct1, 0> : !u16i
+}
```
But removing the second `expected-error` would make the test fail. Is there any way to get around this?

sitio-couto replied:
Oh, my bad. My suggestion is nonsense. The issue is not as trivial as I thought.

The problem is that two errors are being triggered: the invalid attribute type and the types mismatch. It's likely you added the global to be able to create an attribute, but this is not necessary. In these cases, you can use attribute aliases instead.

Fixing my suggestion:

```suggestion
// expected-error@+1 {{type of a DataMemberPtrAttr must be MemberPtrType}}
#test =  #cir.data_member_ptr<!struct1, 0> : !u16i
```

I believe this should work.

Lancern replied:
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExprConst.cpp:
```diff
@@ -1584,6 +1583,30 @@ mlir::Value CIRGenModule::buildNullConstant(QualType T, mlir::Location loc) {
+  return builder.create<mlir::cir::ConstantOp>(
+      loc, cirTy,
+      mlir::cir::DataMemberPtrAttr::get(cirTy, cirClsTy,
+                                        fieldDecl->getFieldIndex()));
```
Updated.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -221,6 +221,11 @@ class CIRGenBuilderTy : public CIRBaseBuilderTy {
     return mlir::cir::TypeInfoAttr::get(anonStruct.getType(), fieldsAttr);
   }
 
+  mlir::cir::DataMemberPtrAttr getDataMemberPtr(mlir::cir::MemberPtrType ty,
```
Updated.

Lancern commented:
> Seems that I missed the null pointer value support for ptr-to-data-member. I'll check this and update the PR.

Just added support for null ptr-to-data-member values.

Lancern commented:
Rebased onto the latest main.

sitio-couto reviewed (commented):
@Lancern a few more suggestions and some minor nitpicks.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -2710,12 +2739,12 @@ def CIR_InlineAsmOp : CIR_Op<"asm", [RecursiveMemoryEffects]> {
+    `(`$asm_flavor`,` `{` $asm_string $constraints `}` `)` attr-dict
     `:` type($res)
-   }];  
+   }];
```
My editor is removing these trailing whitespace again. I'll remove all the whitespace-only changes.

Lancern replied:
Updated.

Lancern commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -366,12 +378,17 @@ DataMemberPtrAttr::verify(function_ref<InFlightDiagnostic()> emitError, Type ty,
-  auto memberIndexValue = memberIndex.getValue().getZExtValue();
-  if (memberIndexValue >= clsStructTy.getNumElements()) {
+  auto memberIndexValue = memberIndex.getValue().getSExtValue();
+  if (memberIndexValue >= static_cast<int64_t>(clsStructTy.getNumElements())) {
```
Good catch! Yes we should also check this to prevent someone from mistakenly gives us an incomplete class. In normal CIRGen clang Sema will check this for us, though.

bcardosolopes replied:
But the verifier is also useful to catch bad transformations, so @sitio-couto comments is still valid. Can you please go over all conversations and update with either "applied" or an explanation on why they are not tackled?

Lancern replied:
Updated.

bcardosolopes commented:
Addressing @Lancern comments:

> Thanks for your comment! But TBH I'm sorry I don't get your point exactly.

No problem, let me clarify!

> I don't clearly see how this is possible. A member pointer is totally different from an ordinary pointer. If we insist the reuse of the existing types instead of the new types, one possible way is to use the types compatible with the underlying ABI layout to represent a ptr-to-data-member.

I think my point can be easier to understand if I ask you a different question: how do you plan to lower this to LLVM? CIR already has most of the things you need to have a lower level representation to handle this just like LLVM does. 

> On most platforms, a ptr-to-data-member is simply represented by an integer which represents the offset of the target field from the beginning of the containing object. So we might use a !s32i or a !s64i to represent a ptr-to-data-member. But this would make a ptr-to-data-member indistinguishable from a normal integer in the CIR, which might affect analysis that is interested on this.

Can you give me an example of such analysis and a specific use case you care about? Note that I'm not against adding nice abstractions, I'm all for it, but I rather not introduce complexity for things we don't have concrete use cases in mind. Note that if we add the abstractions like you are proposing, you'd also need to lower it down to more lower level CIR as part of the LoweringPrepare pass *before* we go down to LLVM. Without some roadmap on how you'd like to use this abstractions, I rather they only get added upon compromise on some upcoming feature that's going to use it.

> Another possible way to represent the ptr-to-member type in CIR is to invent an imagined cir.member type and then cir.ptr<cir.member> represents a ptr-to-member. Are you suggesting this method? This method diverges from the C++ language since the language does not have a corresponding concept.

I'm looking at https://godbolt.org/z/hjaMcqKov and CIR has all the operations necessary to the same approach, can you elaborate on the specific limitations you've found?

bcardosolopes commented:
Addressing @sitio-couto 

> This might work, but we would be losing source code information and diverging from the original codegen. These two types are inherently different, as seen in the example above. Having separate types should facilitate both analysis and lowering.

I'm willing to bite that compromise unless there's a concrete use case in mind. So far, lowering to LLVM sounds like the only one.

> It might also come in handy when we have a proper solution for dealing with target/ABI-specific stuff.

Lowering these ABI details early sounds fine to me, how much these featues get used in practice to justify the abstraction? I don't have numbers nor am I asking for them, but I don't stumble into data-member pointers that often, I might be biased tho.

> I believe this is necessary regardless of the !cir.member_ptr. My understanding is that the get_indirect_member uses variable indexes, while get_member only allows constant indexes. This is essential for pointer-to-member types as they might be unknown at compile time (e.g. when passed as function arguments as shown in the example). We could merge these two operations into a more generic get_member that can do both, but the separation might be better as an initial step.

It's a good way to fix it while keeping nice abstraction, but I'm not yet convinced there are users to take advantage of the benefits.

sitio-couto commented:
@bcardosolopes 

> lowering to LLVM sounds like the only one.

IMO, this and being semantically closer to C/C++ is enough to at least justify the type and the attribute. After all, the complexity of these two components is minimal.

> how much these featues get used in practice to justify the abstraction?

I do not know, but this may also be a reason for us to be conservative by adding said abstraction.

> I'm not yet convinced there are users

I don't think there are. GEP's with runtime indexes are not uncommon in the [original codegen](https://godbolt.org/z/ooczPPxY3). In CIR, however, most are replaced with pointer strides ([example](https://godbolt.org/z/bvnbffWMb)). We could go the same route with data-member pointers.

Lancern commented:
@bcardosolopes 

> Can you give me an example of such analysis and a specific use case you care about?

Unfortunately I cannot give a concrete example at this moment that shows the immediate usefulness of the newly added stuff in this PR. But I believe that it brings benefits similar to the `cir.get_member` operation.

> Without some roadmap on how you'd like to use this abstractions, I rather they only get added upon compromise on some upcoming feature that's going to use it.

This makes sense to me. So if it finally comes to a consenus that this PR is too early to be landed at this stage of this project, I'll close this PR and draft a new PR that directly implements ptr-to-data-member upon existing CIR types and operations.

bcardosolopes commented:
> This makes sense to me. So if it finally comes to a consenus that this PR is too early to be landed at this stage of this project, I'll close this PR and draft a new PR that directly implements ptr-to-data-member upon existing CIR types and operations.

I'm still open to discuss if you feel very strong about this. For instance, you initially suggest to add `!cir.member_ptr`, 
`#cir.data_member_ptr` and `cir.get_indirect_member`, but I didn't see any argument into the direction of a more minimal approach that still keeps some of the semantics you care. It doesn't mean this should be all in or all out. Alternatively, you could do the directly implementation (because that unlocks the fast path to LLVM) and do later incremental work on the pieces you want to raise.

Lancern commented:
Rebased onto the latest `main`.

bcardosolopes commented:
Apologies for the delay here. I finally had some time to take a closer look at this. Before you make changes to this patch, let's perhaps wrap this discussion and converge on a resolution first, so you don't waste resources. My comments/suggestions follow:

# case 1
```
int Point::*pt_member = &Point::x;
// cir.global external @pt_member = #cir.data_member_ptr<!ty_22Point22, 0> : !cir.member_ptr<!s32i in !ty_22Point22>
```
If the type of `!cir.member_ptr` is `!s32i in !ty_22Point22`, it's not really providing much information, given the type doesn't encode any specific information about the element index (for instance, `!ty_22Point22` has 3 ints inside it, which one is this referring to?). Not saying you should encode the information here, just making a point about the fact that it's not really providing much more.

My suggestion:
1. Change `#cir.data_member_ptr<!ty_22Point22, 0>` to `#cir.data_member_ptr<!ty_22Point22[0]>`.
2. Result decay to the actual member pointer type. E.g. `!cir.member_ptr<!s32i in !ty_22Point22>` -> `!cir.ptr<!s32i>`
```
int Point::*pt_member = &Point::x;
// cir.global external @pt_member = #cir.data_member_ptr<!ty_22Point22[0]> : !cir.ptr<!s32i>
```

# case 2
```
auto test1() -> int Point::* {
  return &Point::y;
}
// cir.func @_Z5test1v() -> !cir.member_ptr<!s32i in !ty_22Point22>
//   %{{.+}} = cir.const(#cir.data_member_ptr<!ty_22Point22, 1> : !cir.member_ptr<!s32i in !ty_22Point22>) : !cir.member_ptr<!s32i in !ty_22Point22>
// }
```
My suggestion:
```
auto test1() -> int Point::* {
  return &Point::y;
}
// cir.func @_Z5test1v() -> !cir.ptr<!s32i>
//   %{{.+}} = cir.const(#cir.data_member_ptr<!ty_22Point22[1]> : !cir.ptr<!s32i>) : !cir.ptr<!s32i>
// }
```

# case 3
```
int test2(const Point &pt, int Point::*member) {
  return pt.*member;
}
// cir.func @_Z5test2RK5PointMS_i
//   %{{.+}} = cir.get_indirect_member %{{.+}}[%{{.+}} : !cir.member_ptr<!s32i in !ty_22Point22>] : !cir.ptr<!ty_22Point22> -> !cir.ptr<!s32i>
// }
```
Ok, I think we can still use this one, with some changes:

My suggestion:
1. Rename to `cir.get_runtime_member`.
2. Create an interface `MemberOpInterface` to be shared between `cir.get_member` and `cir.get_runtime_member`. Both ops should have `getAddrTy()` and `getResultTy()` and a new one `isDynamic()`. The idea here is that passes that operate on `cir.get_member` should use the interface instead (and `isDynamic()` is available to distinguish properties).
3. Alternatively, `get_member` could also be changed to support a dynmamic case, but maybe start with an interface is better (@sitio-couto, wdyt?)
```
// cir.func @_Z5test2RK5PointMS_i(%arg0..., %arg1 : !u64i)
//   = cir.get_runtime_member %0[%val : u64i] : (!cir.ptr<!ty_22Point22>) -> !cir.ptr<!s32i>
// }
```

Seems like the same applies to the incomplete types as well. On the `!u64i` choice, we could use datalayout queries to decide between a `!u32i` or `!u64i`. I don't know offhand if `CGF.SizeTy` is what is used for LLVM, but you should that or some similar ABI based type (LLVM codegen should have that information for you to use as inspiration).

Does this makes to you?

Lancern commented:
@bcardosolopes Thanks for your comment! Overall the suggestions make sense, but one problem in case 1 and case 2:

> Result decay to the actual member pointer type. E.g. `!cir.member_ptr<!s32i in !ty_22Point22>` -> `!cir.ptr<!s32i>`

This decay does not make sense because member pointers are not pointers as represented by `!cir.ptr`. Data member pointers are just offsets. In case 3 you lower the `member` parameter to a `!u64i` and that makes more sense. So a more reasonable decay would be `!cir.member_ptr<!s32i in !ty_22Point22>` -> `!u64i`. (p.s. as you said we could use datalayout queries to decide between a `!u32i` or `!u64i`).

For case 1 a more reasonable lowering:

```cpp
int Point::*pt_member = &Point::x;
// cir.global external @pt_member = #cir.data_member_ptr<!ty_22Point22[0]> : !u64i
```

For case 2:

```cpp
auto test1() -> int Point::* {
  return &Point::y;
}
// cir.func @_Z5test1v() -> !u64i
//   %{{.+}} = cir.const(#cir.data_member_ptr<!ty_22Point22[1]> : !u64i) : !u64i
// }
```

Case 3 makes sense to me.

bcardosolopes commented:
> In case 3 you lower the `member` parameter to a `!u64i` and that makes more sense. So a more reasonable decay would be `!cir.member_ptr<!s32i in !ty_22Point22>` -> `!u64i`

Good catch, sounds great to me!

sitio-couto commented:
@bcardosolopes 

> Change `#cir.data_member_ptr<!ty_22Point22, 0>` to `#cir.data_member_ptr<!ty_22Point22[0]>`

This representation will complicate printing/parsing and type-checking. I [suggested](https://github.com/llvm/clangir/pull/401#discussion_r1460614405) `#cir.data_member_ptr<0> : #cir.data_member_ptr<!s32i in !ty_22Point22>` to reuse the type printer/parser: `0` is the field index, `!s32i` the field type, and `!ty_22Point22` the class where to look for the given index. 

> Result decay to the actual member pointer type

> In case 3 you lower the member parameter to a !u64i and that makes more sense.

Is the intent to drop the data-member-pointers type info right after the declaration? If this is the case, the new type will be useless, I think, since it will only be used to declare such data and be dropped right after.

Both comments above apply to Case 2 as well.

> but maybe start with an interface is better

An interface seems excessive for a temporary solution. Are there any current places where said `MemberOpInterface` would be used? If not, regular `llvm::isa<Ty>` should suffice where it is needed, otherwise, seems like generalizing 'cir.get_member' would be more productive. LLVM Dialect's GEP op can serve as an example.

bcardosolopes commented:
> This representation will complicate printing/parsing and type-checking. I [suggested](https://github.com/llvm/clangir/pull/401#discussion_r1460614405) `#cir.data_member_ptr<0> : #cir.data_member_ptr<!s32i in !ty_22Point22>` to reuse the type printer/parser: `0` is the field index, `!s32i` the field type, and `!ty_22Point22` the class where to look for the given index.

Works for me.

> Is the intent to drop the data-member-pointers type info right after the declaration? If this is the case, the new type will be useless, I think, since it will only be used to declare such data and be dropped right after. Other codegen sections for data-member-pointer will likely also require change, such as assignments.

One IPO opportunity I can see in future is tracking a perhaps offset on a callsite param and specialize the callee to transform a `get_runtime_member` into a regular `get_member`. This is why I believe it's important to keep the new constant data attribute to represent the offset - however, thinking more about it now, we don't even need that, all we really need is to look into `get_runtime_member`, which already has the struct type, so if we can prove that the dynamic size is in fact a constant, such optimization would still be possible. So you are right, we don't even need the attribute. One thing we need though: make sure that the offset doesn't consider the underlying type size, and only tracks the member position instead.

> An interface seems excessive for a temporary solution. Are there any current places where said `MemberOpInterface` would be used? If not, regular `llvm::isa<Ty>` should suffice where it is needed, otherwise, seems like generalizing 'cir.get_member' would be more productive. LLVM Dialect's GEP op can serve as an example.

I'm in favor of generalizing `cir.get_member`, I just wanted to make sure you didn't have extra concerns.

sitio-couto commented:
@bcardosolopes 

> So you are right, we don't even need the attribute.

My point is not whether we need the new attribute and type. Since we are introducing them, it seems contradictory that the first action we take is to discard this information by decaying it to an integer, especially if it will complicate other areas such as codegen. Considering this, what is the benefit of this decay? The IPO mentioned seems viable regardless of this decay.

> I'm in favor of generalizing `cir.get_member`, I just wanted to make sure you didn't have extra concerns.

I prefer leaving them separate. `get_runtime_member`'s only users are data-member-pointer accesses. Generalizing `get_member` will considerably complicate it to add functionality required by a single client that might be deprecated in the future. Regardless, between the proposed interface and generalizing `get_member`, I do prefer the latter.

BTW, I [mentioned before](https://github.com/llvm/clangir/pull/401#issuecomment-1909993309) that runtime GEP operations are represented as low-level pointer strides and loads in CIR. **If this is by design**, perhaps we should do the same thing with data-member-pointers.

bcardosolopes commented:
Ok, almost converging :) 

> My point is not whether we need the new attribute and type. Since we are introducing them, it seems contradictory that the first action we take is to discard this information by decaying it to an integer

It seems to me that we can easily retrieve information (versus tree matching a bunch of conditions) by having a data attribute (there's no other way to initialze those otherwise) and a dynamic `get_member`. The specific type isn't useless and could have some nice effects (like skipping function arguments during static analysis), but I'm failing to see a real value.

> especially if it will complicate other areas such as codegen.

Can you elaborate on how it will complicate codegen?

@sitio-couto @Lancern: if we had such a type, how do you plan to later lower it to an actual offset in LowerPrepare? Will we need AST information or does CIR already has everything? I'm asking because the current proposed type seems to be: "there's a `cir.int` in the relevant struct", but it doesn't seem to contain any offset info into that struct (which could have many `cir.int`s inside).

>  Considering this, what is the benefit of this decay? The IPO mentioned seems viable regardless.

I have the reverse question, what's the benefit of keeping it? The only ones I see are mentioned above: defer ABI decision when it's a function argument.

@sitio-couto: if you feel strong about this, I'm willing to take it, I probably just need some comments on the questions above, so I know what's the plan of record.

> I would prefer to leave them separate. `get_runtime_member`'s only users are data-member-pointer accesses. Generalizing `get_member` will considerably complicate it to add functionality required by a single client that might be deprecated in the future. However, between the proposed interface and generalizing `get_member`, I do prefer the latter.

I'm fine with having them separate, I don't have a strong take here, tradeoffs seem fair. My idea with the interface was considering that the thing was actually a pointer type, and if that was the case, it'd be silly to have more than one type that behaves like a pointer without an interface to unify queries. However, since the thing represents an offset, it doesn't make sense to use an interface - I forgot to relay that info in previous comments.

> BTW, I [mentioned before](https://github.com/llvm/clangir/pull/401#issuecomment-1909993309) that runtime GEP operations are represented as low-level pointer strides and loads in CIR. **If this is by design**, perhaps we should do the same thing with data-member-pointers.

I don't like pointer strides here because it loses the "this is dynamically walking a struct" semantic, and makes it semantically easy to transform into static in the future. Also, LowerPrepare could lower `get_runtime_member` to a stride, not a problem there.

On another note, we haven't been using strides for struct related code, at least not from what I can currently see in tests. If we want to unify that somehow, should probably be its own PR proposal.

Lancern commented:
@sitio-couto @bcardosolopes Thanks for your discussion and update! Sorry for the delay as I'm on holiday recently. I'm fine with the decision and I have updated this PR according to the reviews. Please have another round of review when you have time :)

I have squashed the commits and rebased it on the latest `main`.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,104 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+}
+
+void DataMemberAttr::print(AsmPrinter &printer) const {
+  printer << '<';
```
Both parser and printer seem simple enough that they could be written directly in tablegen using declarative syntax, did you run into any problems while trying that approach?

Lancern replied:
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -224,6 +224,36 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+  let summary = "Holds a constant data member pointer value";
+  let parameters = (ins AttributeSelfTypeParameter<"">:$type,
+                        "Type":$clsTy,
+                        "IntegerAttr":$memberIndex);
```
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -224,6 +224,36 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+  let description = [{
+    A data member pointer attribute is a literal attribute that represents a
+    constant pointer-to-data-member value.
+  }];
```
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -224,6 +224,36 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+      return $_get(type.getContext(), type, clsTy,
+          IntegerAttr::get(IndexType::get(type.getContext()), memberIndex));
+    }]>,
+  ];
```
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -224,6 +224,36 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+    }]>,
+  ];
+
+  let hasCustomAssemblyFormat = 1;
```
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -111,6 +111,36 @@ def CIR_PointerType : CIR_Type<"Pointer", "ptr",
+    `cir.member_ptr` models the pointer-to-member type in C++.
+  }];
+
+  let parameters = (ins "mlir::Type":$memberTy, "mlir::Type":$clsTy);
```
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -111,6 +111,36 @@ def CIR_PointerType : CIR_Type<"Pointer", "ptr",
+
+  let extraClassDeclaration = [{
+    /// Determine whether this type represents a pointer-to-member-data type.
+    bool isMemberDataPtr() const;
```
Updated. This problem should no longer exist because now `DataMemberType` represents ptr-to-data-member only.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -111,6 +111,36 @@ def CIR_PointerType : CIR_Type<"Pointer", "ptr",
+
+    /// Determine whether this type represents a pointer-to-member-function
+    /// type.
+    bool isMemberFunctionPtr() const;
```
Updated. This problem should no longer exist because now `DataMemberType` represents ptr-to-data-member only.

Lancern commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -1422,8 +1422,19 @@ mlir::Value ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
+
+    auto Ty = CGF.getCIRType(DestTy).cast<mlir::cir::MemberPtrType>();
+    return Builder.create<mlir::cir::ConstantOp>(
+        CGF.getLoc(E->getExprLoc()), Ty, Builder.getNullDataMemberPtrAttr(Ty));
```
Updated.

Lancern commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -248,8 +248,20 @@ def DataMemberPtrAttr : CIR_Attr<"DataMemberPtr", "data_member_ptr",
       return $_get(type.getContext(), type, clsTy,
           IntegerAttr::get(IndexType::get(type.getContext()), memberIndex));
     }]>,
+    AttrBuilderWithInferredContext<(ins "Type":$type), [{
```
Updated.

Lancern commented:
Sorry I missed the old reviews... I have reviewed all those reviews and should have resolved all of them.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -237,6 +237,52 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+    }
+
+    /// Get the type of the class that contains the target member.
+    mlir::Type getClsTy() const {
```
This method doesn't seem to be necessary, or used throughout this PR, please remove it.

Lancern replied:
Updated.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIROps.td:
```diff
@@ -1774,6 +1774,35 @@ def GetMemberOp : CIR_Op<"get_member"> {
+    The `cir.get_runtime_member` operation gets the address of a member from
+    the input record. The target member is given by a value of type
+    `!cir.data_member` (i.e. a pointer-to-data-member value).
+
```
Please mention in the docs how this differs from `GetMemberOp`.

Lancern replied:
Updated.

bcardosolopes replied:
wow, awesome job, thanks!

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenBuilder.h:
```diff
@@ -865,6 +881,19 @@ class CIRGenBuilderTy : public CIRBaseBuilderTy {
+    auto memberPtrTy = memberPtr.getType().cast<mlir::cir::DataMemberType>();
+
+    // TODO(cir): consider address space.
+    assert(!UnimplementedFeature::addressSpace());
```
Nice!

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,47 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+                       std::optional<size_t> memberIndex) {
+  auto clsStructTy = ty.getClsTy();
+  if (!clsStructTy) {
+    emitError() << "class type of a DataMemberAttr must be a struct type";
```
My understanding from this PR is that a struct type is required when you create a `DataMemberType` so this will never be executed? Either add a testcase that exercises this or remove it.

Lancern replied:
Updated. Originally the code is

```cpp
auto clsStructTy = ty.getClsTy().dyn_cast<StructType>();
if (!clsStructTy) {
  // ...
}
```

I just forgot to remove the `if` block after I removed the `dyn_cast` call.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,47 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+
+  if (clsStructTy.isIncomplete()) {
+    emitError() << "trying to build a non-null ptr to data member of an "
+                   "incomplete class";
```
this should probably read like `incomplete 'cir.struct' requires a non-null data member pointer`.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,47 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+
+  auto memberIndexValue = memberIndex.value();
+  if (memberIndexValue >= clsStructTy.getNumElements()) {
+    emitError() << "member index of a DataMemberAttr is out of range";
```
`DataMemberAttr` -> `#cir.data_member`

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,47 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+  auto memberTy = clsStructTy.getMembers()[memberIndexValue];
+  if (memberTy != ty.getMemberTy()) {
+    emitError()
+        << "member type of a DataMemberAttr must match the attribute type";
```
`DataMemberAttr` -> `#cir.data_member`

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -403,6 +403,30 @@ uint64_t PointerType::getPreferredAlignment(
+llvm::TypeSize
+DataMemberType::getTypeSizeInBits(const ::mlir::DataLayout &dataLayout,
+                                  ::mlir::DataLayoutEntryListRef params) const {
+  // FIXME: consider size differences under different ABIs
```
Please make these two FIXME's into their own Unimplemented features and add asserts here.

Lancern replied:
The header file `UnimplementedFeatureGuarding.h` is located under `CIR/CodeGen` and this file is located under `CIR/Dialect`. I'm not sure whether it is a good idea to add a guard in the `UnimplementedFeature` class and use it here? Maybe we should move the `UnimplementedFeatureGuarding.h` header to a more common place in its parent directories?

bcardosolopes replied:
Good catch! We probably need a separate feature guarding header for dialect stuff, so we could have one for CIRGen and another for CIR dialect (maybe use a slightly different name for the header?). Anyways, content for a follow up PR (if you want to work on that, great).

Lancern replied:
Sounds good. I'll make a follow up PR on this.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRDialect.cpp:
```diff
@@ -2555,6 +2561,32 @@ LogicalResult GetMemberOp::verify() {
+LogicalResult GetRuntimeMemberOp::verify() {
+  auto recordTy = getAddr().getType().getPointee().dyn_cast<StructType>();
+  if (!recordTy) {
+    emitError() << "expected pointer to a record type";
```
This should be enforced directly in tablegen by requiring the pointer to be a "pointer to a struct", see `ExceptionInfoPtr` for an example.

Lancern replied:
Updated.

bcardosolopes commented on clang/test/CIR/IR/invalid.cir:
```diff
@@ -792,6 +792,76 @@ cir.func @const_type_mismatch() -> () {
 
 // -----
 
+!u16i = !cir.int<u, 16>
```
This is awesome, thanks for adding lots of invalid checks. Can you please add a new `.cir` file with a simple use case of all introduced op/type/attr that tests the parser functionality?

Lancern replied:
Updated.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
```diff
@@ -237,6 +237,47 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+
+  let extraClassDeclaration = [{
+    /// Determine whether the data member pointer is a null pointer value.
+    bool isNullValue() const {
```
This one isn't used anywhere in the PR, either use it in the verifier (see my comment there) or remove it!

Lancern replied:
Updated.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRAttrs.cpp:
```diff
@@ -295,6 +295,43 @@ LogicalResult IntAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+DataMemberAttr::verify(function_ref<InFlightDiagnostic()> emitError,
+                       mlir::cir::DataMemberType ty,
+                       std::optional<size_t> memberIndex) {
+  if (!memberIndex.has_value()) {
```
use `isNullValue()` directly here instead?

Lancern replied:
Updated. I decided to remove this function. Testing `memberIndex.has_value()` seems clearer.

bcardosolopes commented on clang/include/clang/CIR/Dialect/IR/CIRAttrs.td:
````diff
@@ -237,6 +237,47 @@ def ConstPtrAttr : CIR_Attr<"ConstPtr", "ptr", [TypedAttrInterface]> {
+
+    Example:
+    ```
+    #ptr = #cir.data_member<1> : !cir.data_member<!s32i in !ty_22Point22>
````
I see an example of this one in `invalid.cir` but I don't see an example where we exercise the parsing and actually use it - perhaps add an usage instance into the recently added `data-member-ptr.cir`?

Lancern replied:
Updated.

bcardosolopes reviewed (changes requested):
Great, one minor remaining comment and we're finally good, thanks for you patience

Lancern commented:
Rebased onto the latest `main`.

bcardosolopes reviewed (approved):
yay, LGTM

orbiri commented on clang/test/CIR/IR/data-member-ptr.cir:
```diff
@@ -0,0 +1,46 @@
+!s32i = !cir.int<s, 32>
+!ty_22Foo22 = !cir.struct<struct "Foo" {!s32i}>
+
+#global_ptr = #cir.data_member<0> : !cir.data_member<!s32i in !ty_22Foo22>
```
Hey, I've stumbled upon this test when refactoring the `cir.const` assembly.
Can you explain what you were trying to test here? The reason I'm asking is that what you have here is an attribute alias, not a global value. That's the reason that the output MLIR is "inlining" it into:

```
// CHECK-NEXT:   cir.func @get_global_member(%arg0: !cir.ptr<!ty_22Foo22>) {
// CHECK-NEXT:     %0 = cir.const(#cir.data_member<0> : !cir.data_member<!s32i in !ty_22Foo22>) : !cir.data_member<!s32i in !ty_22Foo22>
// CHECK-NEXT:     %1 = cir.get_runtime_member %arg0[%0 : !cir.data_member<!s32i in !ty_22Foo22>] : !cir.ptr<!ty_22Foo22> -> !cir.ptr<!s32i>
// CHECK-NEXT:     cir.return
```

bcardosolopes replied:
@Lancern ^

Lancern replied:
@orbiri Well this should just be an effort to save some typings when writing this test. But clearly it does little help, so you can change it if it's convenient for you.

orbiri replied:
> @orbiri Well this should just be an effort to save some typings when writing this test. But clearly it does little help, so you can change it if it's convenient for you.

From the test name it seems like you tried to test the incorporation of some global property here, but if not - then `get_global_member` is 1:1 equivalent to  `get_runtime_member`, and I'd suggest to remove it.

If indeed you intended to test some global property, I'd suggest to revisit this test to make sure that it won't break over time 🙏

Lancern replied:
> From the test name it seems like you tried to test the incorporation of some global property here, but if not - then get_global_member is 1:1 equivalent to get_runtime_member, and I'd suggest to remove it.

I checked this again and I'm sure they're 1:1 equivalent as you said. I'll remove it in an NFC PR if you request so.

Lancern replied:
> I'll remove it in an NFC PR if you request so.

See #582 .
